File: node/test/events.search.test.ts
import { describe, it, expect } from "vitest";
import request from "supertest";
import { createApp } from "../src/app.js";
import { pool } from "../src/db/pool.js";
import { createTenant, createEndpoint, createEvent } from "./fixtures.js";

async function linkDelivery(eventId: string, endpointId: string): Promise<void> {
  await pool.query("INSERT INTO deliveries (event_id, endpoint_id) VALUES ($1, $2)", [eventId, endpointId]);
}

describe("GET /events", () => {
  it("lists the tenant's events, newest first", async () => {
    const { apiKey, id: tenantId } = await createTenant();
    const older = await createEvent(tenantId, { createdAt: new Date(Date.now() - 60_000) });
    const newer = await createEvent(tenantId, { createdAt: new Date() });
    const app = createApp();

    const res = await request(app).get("/events").set("Authorization", `Bearer ${apiKey}`);

    expect(res.status).toBe(200);
    expect(res.body.events.map((e: { id: string }) => e.id)).toEqual([newer.id, older.id]);
  });

  it("never returns another tenant's events", async () => {
    const { apiKey } = await createTenant();
    const { id: otherTenantId } = await createTenant("other-tenant");
    await createEvent(otherTenantId);
    const app = createApp();

    const res = await request(app).get("/events").set("Authorization", `Bearer ${apiKey}`);

    expect(res.body.events).toEqual([]);
  });

  it("filters by type", async () => {
    const { apiKey, id: tenantId } = await createTenant();
    const shipped = await createEvent(tenantId, { type: "order.shipped" });
    await createEvent(tenantId, { type: "order.created" });
    const app = createApp();

    const res = await request(app).get("/events?type=order.shipped").set("Authorization", `Bearer ${apiKey}`);

    expect(res.body.events.map((e: { id: string }) => e.id)).toEqual([shipped.id]);
  });

  it("filters by id", async () => {
    const { apiKey, id: tenantId } = await createTenant();
    const target = await createEvent(tenantId);
    await createEvent(tenantId);
    const app = createApp();

    const res = await request(app).get(`/events?id=${target.id}`).set("Authorization", `Bearer ${apiKey}`);

    expect(res.body.events.map((e: { id: string }) => e.id)).toEqual([target.id]);
  });

  it("filters by from/to time range", async () => {
    const { apiKey, id: tenantId } = await createTenant();
    const inRange = await createEvent(tenantId, { createdAt: new Date("2026-01-15T00:00:00.000Z") });
    await createEvent(tenantId, { createdAt: new Date("2026-02-15T00:00:00.000Z") }); // outside range
    const app = createApp();

    const res = await request(app)
      .get("/events?from=2026-01-01T00:00:00.000Z&to=2026-01-31T00:00:00.000Z")
      .set("Authorization", `Bearer ${apiKey}`);

    expect(res.body.events.map((e: { id: string }) => e.id)).toEqual([inRange.id]);
  });

  it("filters by endpoint_id, matching only events with a delivery to that endpoint", async () => {
    const { apiKey, id: tenantId } = await createTenant();
    const endpointA = await createEndpoint(tenantId, ["order.created"]);
    const endpointB = await createEndpoint(tenantId, ["order.created"]);
    const eventForA = await createEvent(tenantId);
    await linkDelivery(eventForA.id, endpointA.id);
    const eventForB = await createEvent(tenantId);
    await linkDelivery(eventForB.id, endpointB.id);
    const app = createApp();

    const res = await request(app).get(`/events?endpoint_id=${endpointA.id}`).set("Authorization", `Bearer ${apiKey}`);

    expect(res.body.events.map((e: { id: string }) => e.id)).toEqual([eventForA.id]);
  });
});

describe("GET /events/:id", () => {
  it("returns the event with its fan-out deliveries", async () => {
    const { apiKey, id: tenantId } = await createTenant();
    const endpointA = await createEndpoint(tenantId, ["order.created"]);
    const endpointB = await createEndpoint(tenantId, ["order.created"]);
    const event = await createEvent(tenantId, { type: "order.created", payload: { orderId: "abc" } });
    await linkDelivery(event.id, endpointA.id);
    await linkDelivery(event.id, endpointB.id);
    const app = createApp();

    const res = await request(app).get(`/events/${event.id}`).set("Authorization", `Bearer ${apiKey}`);

    expect(res.status).toBe(200);
    expect(res.body).toMatchObject({ id: event.id, type: "order.created", payload: { orderId: "abc" } });
    expect(res.body.deliveries).toHaveLength(2);
    const endpointIds = res.body.deliveries.map((d: { endpoint_id: string }) => d.endpoint_id).sort();
    expect(endpointIds).toEqual([endpointA.id, endpointB.id].sort());
  });

  it("returns 404 for an event belonging to a different tenant", async () => {
    const { id: otherTenantId } = await createTenant("other-tenant");
    const foreignEvent = await createEvent(otherTenantId);
    const { apiKey } = await createTenant();
    const app = createApp();

    const res = await request(app).get(`/events/${foreignEvent.id}`).set("Authorization", `Bearer ${apiKey}`);

    expect(res.status).toBe(404);
  });
});

File: go/internal/api/events_search_test.go
package api_test

import (
	"context"
	"net/http"
	"testing"
	"time"

	"github.com/jackc/pgx/v5/pgxpool"
	"github.com/stretchr/testify/require"

	"webhooks-go/internal/testsupport"
)

// insertDelivery links an already-created event to an endpoint directly —
// bypasses expansion for tests that only need the fan-out relationship to
// exist, not a real expansion cycle to have run.
func insertDelivery(t *testing.T, pool *pgxpool.Pool, eventID, endpointID string) {
	t.Helper()
	_, err := pool.Exec(context.Background(),
		"INSERT INTO deliveries (event_id, endpoint_id) VALUES ($1, $2)", eventID, endpointID,
	)
	require.NoError(t, err)
}

func TestListEventsNewestFirst(t *testing.T) {
	pool := testsupport.SetupPool(t)
	ts := newTestServer(t, pool)
	tenantID, apiKey := testsupport.CreateTenant(t, pool)
	older := time.Now().Add(-60 * time.Second)
	newer := time.Now()
	olderID := testsupport.CreateEvent(t, pool, tenantID, testsupport.EventOptions{CreatedAt: &older})
	newerID := testsupport.CreateEvent(t, pool, tenantID, testsupport.EventOptions{CreatedAt: &newer})

	resp := doRequest(t, http.MethodGet, ts.URL+"/events", apiKey, nil)
	require.Equal(t, http.StatusOK, resp.StatusCode)

	var body struct {
		Events []struct{ ID string } `json:"events"`
	}
	decodeJSON(t, resp, &body)
	require.Len(t, body.Events, 2)
	require.Equal(t, []string{newerID, olderID}, []string{body.Events[0].ID, body.Events[1].ID})
}

func TestListEventsNeverReturnsAnotherTenantsEvents(t *testing.T) {
	pool := testsupport.SetupPool(t)
	ts := newTestServer(t, pool)
	_, apiKey := testsupport.CreateTenant(t, pool)
	otherTenantID, _ := testsupport.CreateTenant(t, pool)
	testsupport.CreateEvent(t, pool, otherTenantID, testsupport.EventOptions{})

	resp := doRequest(t, http.MethodGet, ts.URL+"/events", apiKey, nil)
	require.Equal(t, http.StatusOK, resp.StatusCode)

	var body struct {
		Events []struct{ ID string } `json:"events"`
	}
	decodeJSON(t, resp, &body)
	require.Empty(t, body.Events)
}

func TestListEventsFiltersByType(t *testing.T) {
	pool := testsupport.SetupPool(t)
	ts := newTestServer(t, pool)
	tenantID, apiKey := testsupport.CreateTenant(t, pool)
	shippedID := testsupport.CreateEvent(t, pool, tenantID, testsupport.EventOptions{Type: "order.shipped"})
	testsupport.CreateEvent(t, pool, tenantID, testsupport.EventOptions{Type: "order.created"})

	resp := doRequest(t, http.MethodGet, ts.URL+"/events?type=order.shipped", apiKey, nil)
	require.Equal(t, http.StatusOK, resp.StatusCode)

	var body struct {
		Events []struct{ ID string } `json:"events"`
	}
	decodeJSON(t, resp, &body)
	require.Len(t, body.Events, 1)
	require.Equal(t, shippedID, body.Events[0].ID)
}

func TestListEventsFiltersByID(t *testing.T) {
	pool := testsupport.SetupPool(t)
	ts := newTestServer(t, pool)
	tenantID, apiKey := testsupport.CreateTenant(t, pool)
	targetID := testsupport.CreateEvent(t, pool, tenantID, testsupport.EventOptions{})
	testsupport.CreateEvent(t, pool, tenantID, testsupport.EventOptions{})

	resp := doRequest(t, http.MethodGet, ts.URL+"/events?id="+targetID, apiKey, nil)
	require.Equal(t, http.StatusOK, resp.StatusCode)

	var body struct {
		Events []struct{ ID string } `json:"events"`
	}
	decodeJSON(t, resp, &body)
	require.Len(t, body.Events, 1)
	require.Equal(t, targetID, body.Events[0].ID)
}

func TestListEventsFiltersByFromToRange(t *testing.T) {
	pool := testsupport.SetupPool(t)
	ts := newTestServer(t, pool)
	tenantID, apiKey := testsupport.CreateTenant(t, pool)
	inRange := time.Date(2026, 1, 15, 0, 0, 0, 0, time.UTC)
	outOfRange := time.Date(2026, 2, 15, 0, 0, 0, 0, time.UTC)
	inRangeID := testsupport.CreateEvent(t, pool, tenantID, testsupport.EventOptions{CreatedAt: &inRange})
	testsupport.CreateEvent(t, pool, tenantID, testsupport.EventOptions{CreatedAt: &outOfRange})

	resp := doRequest(t, http.MethodGet, ts.URL+"/events?from=2026-01-01T00:00:00.000Z&to=2026-01-31T00:00:00.000Z", apiKey, nil)
	require.Equal(t, http.StatusOK, resp.StatusCode)

	var body struct {
		Events []struct{ ID string } `json:"events"`
	}
	decodeJSON(t, resp, &body)
	require.Len(t, body.Events, 1)
	require.Equal(t, inRangeID, body.Events[0].ID)
}

func TestListEventsFiltersByEndpointID(t *testing.T) {
	pool := testsupport.SetupPool(t)
	ts := newTestServer(t, pool)
	tenantID, apiKey := testsupport.CreateTenant(t, pool)
	endpointA := testsupport.CreateEndpoint(t, pool, tenantID, []string{"order.created"}, testsupport.EndpointOptions{})
	endpointB := testsupport.CreateEndpoint(t, pool, tenantID, []string{"order.created"}, testsupport.EndpointOptions{})
	eventForA := testsupport.CreateEvent(t, pool, tenantID, testsupport.EventOptions{})
	insertDelivery(t, pool, eventForA, endpointA)
	eventForB := testsupport.CreateEvent(t, pool, tenantID, testsupport.EventOptions{})
	insertDelivery(t, pool, eventForB, endpointB)

	resp := doRequest(t, http.MethodGet, ts.URL+"/events?endpoint_id="+endpointA, apiKey, nil)
	require.Equal(t, http.StatusOK, resp.StatusCode)

	var body struct {
		Events []struct{ ID string } `json:"events"`
	}
	decodeJSON(t, resp, &body)
	require.Len(t, body.Events, 1)
	require.Equal(t, eventForA, body.Events[0].ID)
}

func TestGetEventReturnsFanOutDeliveries(t *testing.T) {
	pool := testsupport.SetupPool(t)
	ts := newTestServer(t, pool)
	tenantID, apiKey := testsupport.CreateTenant(t, pool)
	endpointA := testsupport.CreateEndpoint(t, pool, tenantID, []string{"order.created"}, testsupport.EndpointOptions{})
	endpointB := testsupport.CreateEndpoint(t, pool, tenantID, []string{"order.created"}, testsupport.EndpointOptions{})
	eventID := testsupport.CreateEvent(t, pool, tenantID, testsupport.EventOptions{Type: "order.created", Payload: `{"orderId":"abc"}`})
	insertDelivery(t, pool, eventID, endpointA)
	insertDelivery(t, pool, eventID, endpointB)

	resp := doRequest(t, http.MethodGet, ts.URL+"/events/"+eventID, apiKey, nil)
	require.Equal(t, http.StatusOK, resp.StatusCode)

	var body struct {
		ID         string         `json:"id"`
		Type       string         `json:"type"`
		Payload    map[string]any `json:"payload"`
		Deliveries []struct {
			EndpointID string `json:"endpoint_id"`
		} `json:"deliveries"`
	}
	decodeJSON(t, resp, &body)
	require.Equal(t, eventID, body.ID)
	require.Equal(t, "order.created", body.Type)
	require.Equal(t, "abc", body.Payload["orderId"])
	require.Len(t, body.Deliveries, 2)
	endpointIDs := []string{body.Deliveries[0].EndpointID, body.Deliveries[1].EndpointID}
	require.ElementsMatch(t, []string{endpointA, endpointB}, endpointIDs)
}

func TestGetEventReturnsNotFoundForCrossTenantEvent(t *testing.T) {
	pool := testsupport.SetupPool(t)
	ts := newTestServer(t, pool)
	otherTenantID, _ := testsupport.CreateTenant(t, pool)
	foreignEventID := testsupport.CreateEvent(t, pool, otherTenantID, testsupport.EventOptions{})
	_, apiKey := testsupport.CreateTenant(t, pool)

	resp := doRequest(t, http.MethodGet, ts.URL+"/events/"+foreignEventID, apiKey, nil)
	require.Equal(t, http.StatusNotFound, resp.StatusCode)
}
